ops/model.py

```diff
@@ -796,15 +801,20 @@ def __repr__(self):
 class RelationMapping(Mapping[str, List['Relation']]):
     """Map of relation names to lists of :class:`Relation` instances."""
 
-    def __init__(self, relations_meta: Dict[str, 'ops.RelationMeta'], our_unit: 'Unit',
-                 backend: '_ModelBackend', cache: '_ModelCache'):
+    def __init__(self,
```

Comment:
I'm somewhat surprised we don't need to modify `__iter__` / `__contains__` / `__len__`. Wouldn't these have to exclude the broken relation too, or am I missing something?

I guess we don't need any changes to `Model.get_relation` / `RelationMapping._get_unique` because there they're asking for a specific relation?

Reply:
`_get_unique` changes mostly indirectly.

If the relation ID is passed, then it does `for relation in self[relation_name]` and that utilises the change in `__getitem__` to skip past the broken relation, so we end up in the `else` clause of the `for` loop, which does have the `active=False` change. I feel this is the right behaviour.

If the relation ID is not passed, then it does (essentially) `len(self[relation_name])` which again skips the broken relation. In that case it'll return None rather than a `Relation` with `active==False`. I'm not certain that this is correct.

The docs say "If relation_id is not given, this will return the Relation instance if the relation is established only once or None if it is not established". It seemed like "not established" was functionally the same as "broken", so you'd expect None. Perhaps that's wrong, though?

If you have a broken relation (ID 10, name "db") and do `Model.get_relation("db", 10)` you get a `Relation` with `active==False`, and if you do `Model.get_relation("db")` you get `None`. I think I have just about convinced myself that it should actually give the same in both cases, but let me know what you think.

Reply:
`__iter__`, `__contains__`, and `__len__` I think I missed :disappointed:.

Reply:
> `__iter__`, `__contains__`, and `__len__` I think I missed 😞.

No, I just forgot that I'd looked at them :laughing:.

These all currently work on the defined relations, ie. what's in the metadata, and not the current state.

For example, if I have a charm that has this in `charmcraft.yaml`/`metadata.yaml`:

```yaml
requires:
    database:
        interface: postgresql_client
        limit: 1
```

And I deploy the charm, *but I do not install postgresql and do not do a `juju integrate postgresql-k8s my-charm`*, so there is a relation required but not made, then:

* `len(self.model.relations)` is `1`
* `"database" in self.model.relations` is `True`
* `for rel_name in self.model.relations:` will have one iteration, where `rel_name` is `database`

If I do install postgresql and do `juju integrate` then those all stay the same - it's `self.model.relations["database"]` that will change, from `[]` to `[Relation(...)]`

I don't think anything should change here with relation-broken. If the relation is broken, it's still defined/required/provided.

Reply:
Excellent, thanks for the replies -- resolving.